server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RandomStoring.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.james.transport.mailets.delivery.MailDispatcher;
+import org.apache.james.transport.mailets.delivery.MailboxAppender;
+import org.apache.james.transport.mailets.delivery.SimpleMailStore;
+import org.apache.james.user.api.UsersRepository;
+import org.apache.james.user.api.UsersRepositoryException;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMailet;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.Streams;
+
+/**
+ * Receives a Mail from the Queue and takes care of delivery of the
+ * message to local inboxes.
+ *
+ * This mailet is a composition of RecipientRewriteTable, SieveMailet
+ * and MailboxManager configured to mimic the old "LocalDelivery"
+ * James 2.3 behavior.
+ */
+public class RandomStoring extends GenericMailet {
+
+    private static final String RANDOM_STORING_MAILS_METRIC_NAME = "randomStoringMails";
+    private static final int MIN_NUMBER_OF_RECIPIENTS = 4;
+    private static final int MAX_NUMBER_OF_RECIPIENTS = 8;
+    private final UsersRepository usersRepository;
+    private final MailboxManager mailboxManager;
+    private final MetricFactory metricFactory;
+    private final Iterator<Integer> randomSizes;
+    private MailDispatcher mailDispatcher;
+
+    @Inject
+    public RandomStoring(UsersRepository usersRepository, @Named("mailboxmanager") MailboxManager mailboxManager,
+                         MetricFactory metricFactory) {
+        this.metricFactory = metricFactory;
+        this.usersRepository = usersRepository;
+        this.mailboxManager = mailboxManager;
+        this.randomSizes = new Random().ints(MIN_NUMBER_OF_RECIPIENTS, MAX_NUMBER_OF_RECIPIENTS + 1).boxed().iterator();
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        try {
+            mail.setRecipients(generateRandomUsers());
+            mailDispatcher.dispatch(mail);
+        } catch (UsersRepositoryException e) {
+            throw new MessagingException("Unable to compute a random user list", e);
+        }
+    }
+
+    @Override
+    public String getMailetInfo() {
+        return "Random Storing Mailet";
+    }
+
+    @Override
+    public void init() throws MessagingException {
+        mailDispatcher = MailDispatcher.builder()
+            .mailStore(SimpleMailStore.builder()
+                .mailboxAppender(new MailboxAppender(mailboxManager))
+                .usersRepository(usersRepository)
+                .folder(MailboxConstants.INBOX)
+                .metric(metricFactory.generate(RANDOM_STORING_MAILS_METRIC_NAME))
+                .build())
+            .consume(getInitParameter("consume", true))
+            .mailetContext(getMailetContext())
+            .build();
+    }
+
+    private Collection<MailAddress> generateRandomUsers() throws UsersRepositoryException {
+        List<MailAddress> mailAddresses = Streams.stream(usersRepository
```

Comment:
I would have taken the list of recipients as a configuration parameter to make things simpler